src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -310,54 +311,46 @@ private void WriteAsGnuSharedInternal(Span<byte> buffer, out long actualLength,
         {
-            // The ustar fields (uid, gid, linkName, uname, gname, devmajor, devminor) do not get written.
-            // The mode gets the default value.
-            _name = GenerateExtendedAttributeName();
-            _mode = TarHelpers.GetDefaultMode(_typeFlag);
-            _typeFlag = isGea ? TarEntryType.GlobalExtendedAttributes : TarEntryType.ExtendedAttributes;
-            _linkName = string.Empty;
-            _magic = string.Empty;
-            _version = string.Empty;
-            _gName = string.Empty;
-            _uName = string.Empty;
-
+            WriteAsPaxExtendedAttributesShared(isGea, globalExtendedAttributesEntryNumber);
             _dataStream = GenerateExtendedAttributesDataStream(extendedAttributes);
-
             WriteAsPaxInternal(archiveStream, buffer);
         }
 
         // Asynchronously writes the current header as a PAX Extended Attributes entry into the archive stream and returns the value of the final checksum.
-        private async Task<int> WriteAsPaxExtendedAttributesAsync(Stream archiveStream, Memory<byte> buffer, IEnumerable<KeyValuePair<string, string>> extendedAttributes, bool isGea, CancellationToken cancellationToken)
+        private async Task WriteAsPaxExtendedAttributesAsync(Stream archiveStream, Memory<byte> buffer, Dictionary<string, string> extendedAttributes, bool isGea, int globalExtendedAttributesEntryNumber, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            // The ustar fields (uid, gid, linkName, uname, gname, devmajor, devminor) do not get written.
-            // The mode gets the default value.
-            _name = GenerateExtendedAttributeName();
-            _mode = TarHelpers.GetDefaultMode(_typeFlag);
-            _typeFlag = isGea ? TarEntryType.GlobalExtendedAttributes : TarEntryType.ExtendedAttributes;
-            _linkName = string.Empty;
-            _magic = string.Empty;
-            _version = string.Empty;
-            _gName = string.Empty;
-            _uName = string.Empty;
-
+            WriteAsPaxExtendedAttributesShared(isGea, globalExtendedAttributesEntryNumber);
             _dataStream = await GenerateExtendedAttributesDataStreamAsync(extendedAttributes, cancellationToken).ConfigureAwait(false);
+            await WriteAsPaxInternalAsync(archiveStream, buffer, cancellationToken).ConfigureAwait(false);
+        }
+
+        // Initializes the name, mode and type flag of a PAX extended attributes entry.
+        private void WriteAsPaxExtendedAttributesShared(bool isGea, int globalExtendedAttributesEntryNumber)
+        {
+            Debug.Assert(isGea && globalExtendedAttributesEntryNumber >= 0 || !isGea && globalExtendedAttributesEntryNumber < 0);
 
-            return await WriteAsPaxInternalAsync(archiveStream, buffer, cancellationToken).ConfigureAwait(false);
+            _name = isGea ?
```

Comment:
Now I can write the name of both a G.E.A. entry or a E.A. entry in the same place, instead of at different levels in the stack due to the async problem.